src/react/DevicesPage.jsx

```diff
@@ -35,126 +36,150 @@ import SingleCardPage from './ui/SingleCardPage';
+  const [filters, setFilters] = useState({});
 
-  const handleDevicesRequest = (pageIndex, response) => {
-    const token = new URLSearchParams(response.links.next).get('from_token');
-    const deviceList = response.data.map((value) => Device.fromObject(value));
-    cachedPages[pageIndex] = {
-      devices: deviceList,
-      token,
-    };
-    setCachedPages([...cachedPages]);
-    if (pageIndex === activePage) {
-      setPhase('ok');
-    }
-    const pagesToCache = activePage + MAX_SHOWN_PAGES + 1;
-    if (pageIndex < pagesToCache) {
-      // eslint-disable-next-line no-use-before-define
-      cachePage(pageIndex + 1);
+  const pageAlerts = useAlerts();
+  const pagedDevices = useMemo(() => {
+    const pageList = [];
+    const devices = deviceList.filter((device) => matchFilters(device, filters));
+    while (devices.length > 0) {
+      pageList.push(devices.splice(0, DEVICES_PER_PAGE));
     }
-    return null;
-  };
+    return pageList;
+  }, [deviceList, filters]);
 
-  const cachePage = (pageIndex) => {
-    if (pageIndex === 0) {
-      astarte
-        .getDevices({
-          details: true,
-          limit: DEVICES_PER_PAGE,
-        })
-        .then((response) => handleDevicesRequest(pageIndex, response))
-        .catch((err) =>
-          pageAlerts.showError(`Couldn't retrieve the device list from Astarte: ${err.message}`),
-        );
-    } else {
-      const { token } = cachedPages[pageIndex - 1];
-      if (token) {
-        astarte
-          .getDevices({
-            details: true,
-            from: token,
-            limit: DEVICES_PER_PAGE,
-          })
-          .then((response) => handleDevicesRequest(pageIndex, response))
-          .catch((err) =>
-            pageAlerts.showError(`Couldn't retrieve the device list from Astarte: ${err.message}`),
-          );
-      }
+  const loadMoreDevices = async (fromToken, loadAllDevices = false) =>
+    astarte
+      .getDevices({
+        details: true,
+        from: fromToken,
+        limit: DEVICES_PER_REQUEST,
+      })
+      .then((resp) => {
+        const nextToken = new URLSearchParams(resp.links.next).get('from_token');
+        const devices = resp.data.map((value) => Device.fromObject(value));
+        setRequestToken(nextToken);
+
+        setDeviceList((previousList) => {
+          const updatedDeviceList = previousList.concat(devices);
+
+          const pageCount = Math.ceil(updatedDeviceList.length / DEVICES_PER_PAGE);
+          const shouldLoadMore = pageCount < activePage + MAX_SHOWN_PAGES || loadAllDevices;
+          if (shouldLoadMore && nextToken) {
+            loadMoreDevices(nextToken, loadAllDevices);
+          } else {
+            setPhase('ok');
+          }
+
+          return updatedDeviceList;
+        });
+      })
+      .catch((err) => {
+        pageAlerts.showError(`Couldn't retrieve the device list from Astarte: ${err.message}`);
+      });
+
+  const pageChangeHandler = (pageIndex) => {
```

Comment:
A common pattern for functions handling UI events is to prefix them with `handle`:
`pageChangeHandler` -> `handlePageChange`.
The same goes for `filterUpdateHandler`

Reply:
No strong opinions here too. Since you're defining a constant, eventHandler tells me that this constant contains a function that handles an event. But I do agree it's easier to read `onPageChange={handlePageChange}`

Reply:
Then I would suggest sticking to a pattern, even better if we follow patterns common in React codebases.
- if you are creating a component and exposing event hooks, those props would be on: onClick, onHover, onError
- if you are consuming another component, you want to add handling in response to these events, so you use handle: handleChange, handleClick. Because your job is now to handle some event.

https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-handler-names.md

Reply:
I agree, we could also enable that ESLint check.
(not in this PR or it will throw warning everywhere)